esda/correlogram.py

```diff
@@ -0,0 +1,153 @@
+
+    Returns
+    -------
+    pandas.Series
+        a pandas series with the computed autocorrelation statistic and its simulated p-value
+    """
+    (
+        y,  # y variable
+        tree,  # kd tree
+        W,  # weights class DistanceBand or KNN
+        STATISTIC,  # class of statistic (Moran, Geary, etc)
+        dist,  # threshold/k parameter for the weights
+        weights_kwargs,  # additional args
+        stat_kwargs,  # additional args
+    ) = inputs
+
+    w = W(tree, dist, silence_warnings=True, **weights_kwargs)
+    autocorr = STATISTIC(y, w, **stat_kwargs)
+    attrs = []
+    all_attrs = list(dict(vars(autocorr)).keys())
+    for attribute in all_attrs:
+        attrs.append(getattr(autocorr, str(attribute)))
+    return pd.Series(attrs, index=all_attrs, name=dist)
+
+
+def correlogram(
+    gdf: gpd.GeoDataFrame,
+    variable: str,
+    distances: list,
+    distance_type: str = "band",
+    statistic: str = "I",
+    weights_kwargs: dict = None,
+    stat_kwargs: dict = None,
+    n_jobs: int = -1,
+    backend: str = "loky",
+):
+    """Generate a spatial correlogram
+
+    A spatial correlogram is a set of spatial autocorrelation statistics calculated for
+    a set of increasing distances. It is a useful exploratory tool for examining
+    how the relationship between spatial units changes over different notions of scale.
+
+    Parameters
+    ----------
+    gdf : gpd.GeoDataFrame
+        geodataframe holding spatial and attribute data
+    variable: str
+        column on the geodataframe used to compute autocorrelation statistic
+    distances : list
+        list of distances to compute the autocorrelation statistic
+    distance_type : str, optional
+        which concept of distance to increment. Options are {`band`, `knn`}.
+        by default 'band' (for `libpysal.weights.DistanceBand` weights)
+    statistic : str, by default 'I'
+        which spatial autocorrelation statistic to compute. Options in {`I`, `G`, `C`}
+    weights_kwargs : dict
+        additional keyword arguments passed to the libpysal.weights.W class
+    stat_kwargs : dict
+        additional keyword arguments passed to the `esda` autocorrelation statistic class.
+        For example for faster results with no statistical inference, set the number
+        of permutations to zero with {permutations: 0}
+    n_jobs : int
+        number of jobs to pass to joblib. If -1 (default), all cores will be used
+    backend : str
+        backend parameter passed to joblib
+
+    Returns
+    -------
+    outputs : pandas.DataFrame
+        table of autocorrelation statistics at increasing distance bandwidths
+    """
+    if stat_kwargs is None:
+        stat_kwargs = dict()
+    if weights_kwargs is None:
+        weights_kwargs = dict()
+    if statistic == "I":
+        STATISTIC = Moran
+    elif statistic == "G":
+        STATISTIC = G
+    elif statistic == "C":
+        STATISTIC = Geary
+    else:
```

Comment:
cool, that makes sense. I think the first version I wrote actually took the Moran class as the first argument, but i thought the dispatching actually helped clarify what the function was doing. The internals ended up a lot more generic than i figured they'd need to be

so the most abstract version returns the output of a callable for a range of W/Graph specifications (in table form), and there are probably plenty of uses for that. But then do we want to call the function something more generic since it no longer necessarily computes some measure of correlation?

Reply:
Totally, and then define a specific `correlogram()` function that only supports Moran, spatial Pearson, or spatial tau? 

For the general function, @TaylorOshan and I have been calling these "profile" functions ([bottom of page 302](https://github.com/pysal/esda/files/12262367/s10109-022-00384-8.pdf)), so I'm partial to `distance_profile()` or `spatial_profile()`.

Reply:
Well, I guess even like.... I have a nascent implementation of a nonparametric robust Moran/Local Moran statistic that would benefit from being able to plug in directly? I think it's OK to call the function `correlogram()`, and then allow for a user-defined callable. The fact that it's user-defined means that we can't enforce the output is callable, and this keeps things more simple for the user.

Reply:
profile is the same nomenclature [over in segregation](https://pysal.org/segregation/api.html#dynamics) but maybe we should centralize some of this logic then